packages/components/src/DataList/DataList.utils.tsx

```diff
@@ -0,0 +1,143 @@
+import React, { Children, ReactElement, isValidElement } from "react";
+import isEmpty from "lodash/isEmpty";
+import {
+  DataListHeader,
+  DataListItemType,
+  DataListItemTypeFromHeader,
+  DataListObject,
+} from "./DataList.types";
+import styles from "./DataList.css";
+import { EmptyState, EmptyStateProps } from "./components/EmptyState";
+import {
+  EMPTY_FILTER_RESULTS_ACTION_LABEL,
+  EMPTY_FILTER_RESULTS_MESSAGE,
+} from "./DataList.const";
+import { FormatDate } from "../FormatDate";
+import { InlineLabel } from "../InlineLabel";
+import { Text } from "../Text";
+
+/**
+ * Return the child component that matches the `type` provided
+ */
+export function getCompoundComponent<T>(
+  children: ReactElement | ReactElement[],
+  type: ReactElement<T>["type"],
+): ReactElement<T> | undefined {
+  const childrenArray = Children.toArray(children);
+  const element = childrenArray.find(
+    child => isValidElement<T>(child) && child.type === type,
+  );
+
+  // Comply with the return type without casting it
+  return isValidElement<T>(element) ? element : undefined;
+}
+
+/**
+ * Generate the default elements the DataList would use on the data provided.
+ */
+export function generateElementsFromData<T extends DataListObject>(data: T[]) {
+  type DataListElements = DataListItemType<typeof data>;
+
+  return data.map(item =>
+    Object.keys(item).reduce((acc, key: keyof DataListElements) => {
+      const currentItem = item[key];
+
+      if (!currentItem) {
+        return acc;
+      }
+
+      if (key === "tags" && Array.isArray(currentItem)) {
+        acc[key] = (
+          // TODO: Create a component specific for this with the experience we want
+          <div style={{ display: "flex", gap: 8, overflow: "hidden" }}>
+            {currentItem.filter(Boolean).map((tag, index) => (
+              <InlineLabel key={index}>{tag}</InlineLabel>
+            ))}
+          </div>
+        );
+      } else if (key === "label" && typeof currentItem === "string") {
+        acc[key] = <Text>{currentItem}</Text>;
+      } else if (isValidElement(currentItem)) {
+        acc[key] = currentItem;
+      } else if (currentItem instanceof Date) {
+        acc[key] = (
+          <Text variation="subdued">
+            <FormatDate date={currentItem} />
+          </Text>
+        );
+      } else {
+        acc[key] = <Text variation="subdued">{currentItem}</Text>;
+      }
+
+      return acc;
+    }, {} as DataListElements),
+  );
+}
+
+/**
+ * Generate the header elements with the default styling
+ */
+export function generateHeaderFromData<T extends DataListObject>(
```

Comment:
Super nitpicky, but I am not a big fan of the `FromData` mainly because this is actually from the `headers` prop and the `data` prop is different and isn't used in the `headers` besides for helping with the typescript magic.

I can't think of a better name so if you have nothing I am not going to block it for this

Reply:
hmm yeah, i'll think on it. not crazy about it either

Reply:
@darryltec thoughts on changing the `headers` prop to `columns`?

Reply:
Thought about this more. I think we should defer that back to the sorting story. Mostly because if our main reason is for it to be flexible for when sorting comes, it could also go either way. Following the pattern of "simple props" declaring sort should be on the sort prop like so `sort={{sortable: ["label", "address"], onSort: (sorted) => void}}`